MultimodalQnA/ui/gradio/multimodalqna_ui_gradio.py

```diff
@@ -56,98 +56,123 @@ def clear_history(state, request: gr.Request):
+        if Path(files[0]).suffix in GRADIO_AUDIO_FORMATS or len(files) > 1:
+            audio_file = files[-1]  # Guaranteed that last file would be recorded audio
+    
+    # Add to chatbot history
+    if image_file:
+        state.image_query_file = image_file
+        state.chatbot_history.append({
+            "role": state.roles[0],
+            "content": {"path": image_file}
+        })
+    if audio_file:
+        state.audio_query_file = audio_file
+        state.chatbot_history.append({
+            "role": state.roles[0],
+            "content": {"path": audio_file}
+        })
+    
+    state.chatbot_history.append({
+        "role": state.roles[0],
+        "content": text
+    }) 
+    
+    logger.info(f"add_text. ip: {request.client.host}. len: {len(text)}")
+    
+    return (
+        state,
+        state.to_gradio_chatbot(),
+        gr.MultimodalTextbox(value=None)
+    ) + (disable_btn,) * 1
 
 
-def http_bot(state, request: gr.Request):
+def http_bot(state, audio_response_toggler, request: gr.Request):
     global gateway_addr
     logger.info(f"http_bot. ip: {request.client.host}")
     url = gateway_addr
-    is_very_first_query = False
-    is_audio_query = state.audio_query_file is not None
+    
     if state.skip_next:
         # This generate call is skipped due to invalid inputs
         yield (state, state.to_gradio_chatbot(), None, None, None) + (no_change_btn,) * 1
         return
-
-    if len(state.messages) == state.offset + 2:
-        # First round of conversation
-        is_very_first_query = True
-        new_state = multimodalqna_conv.copy()
-        new_state.append_message(new_state.roles[0], state.messages[-2][1])
-        new_state.append_message(new_state.roles[1], None)
-        new_state.audio_query_file = state.audio_query_file
-        new_state.image_query_files = state.image_query_files
-        state = new_state
-
+    
+    is_very_first_query = all(True if h['role'] == 'user' else False for h in state.chatbot_history)
+    
     # Construct prompt
-    prompt = state.get_prompt()
+    prompt = state.get_prompt(is_very_first_query)
+    
+    modalities = ["text", "audio"] if audio_response_toggler else ["text"]
 
     # Make requests
     pload = {
         "messages": prompt,
+        "modalities": modalities
     }
-
+    
+    state.chatbot_history.append({
+        "role": state.roles[1],
+        "content": "▌"
+    })
+    
+    yield (state, state.to_gradio_chatbot(), state.split_video, state.image, state.pdf) + (disable_btn,) * 1
+    
     if logflag:
         logger.info(f"==== request ====\n{pload}")
     logger.info(f"==== url request ====\n{gateway_addr}")
 
-    state.messages[-1][-1] = "▌"
-
-    yield (state, state.to_gradio_chatbot(), state.split_video, state.image, state.pdf) + (disable_btn,) * 1
-
     try:
         response = requests.post(
             url,
             headers=headers,
             json=pload,
             timeout=ui_timeout,
         )
+        
```

Comment:
There's some extra white space here and on line 161 that can be removed so that we don't have unnecessary diffs